### PR TITLE
Prep for MPI PA integration

### PR DESCRIPTION
**Pull request by tgerdesnv**
Update ResultManager to store tuple using RunConfig instead of ModelConfig

RunConfig new functions:
- cpu_only()
- models_name()     : for base model names
- model_variants_name()  : for model variants
- from_dict()

ModelRunConfig new functions:
- model_variant_name()
- from_dict()

Updated ModelManager.run_model() to ModelManager.run_models(), which can handle single or multiple concurrent models
Refactor PerfAnalyzer to prep for MPI changes
Update MetricsManager.execute_run_config() and MetricsManager.profile_models() to handle multi-model

### CODE REVIEW

File: model_analyzer/analyzer.py [hunk 1]
```diff
@@ -203,6 +170,54 @@ def report(self, mode):
         self._report_manager.create_detailed_reports()
         self._report_manager.export_detailed_reports()
 
+    def _profile_models(self):
```
Comment (nv-braf) [inline model_analyzer/analyzer.py:173]: Should change the order of these to match how they are called (reads better)

File: model_analyzer/analyzer.py [hunk 2]
```diff
@@ -89,43 +89,10 @@ def profile(self, client, gpus):
...
                 " got {type(self._config)}.")
 
-        self._metrics_manager = MetricsManager(
-            config=self._config,
-            client=client,
-            server=self._server,
-            gpus=gpus,
-            result_manager=self._result_manager,
-            state_manager=self._state_manager)
+        self._create_profile_managers(client, gpus)
```
Comment (nv-braf) [inline model_analyzer/analyzer.py:92]: Can we refactor this further into _create_metrics_manager & _create_model_manager?

File: model_analyzer/perf_analyzer/perf_analyzer.py [hunk 3]
```diff
@@ -131,63 +136,78 @@ def run(self, metrics, env=None):
...
                         "but no valid requests recorded over max request count"
-                        f" of {self._config['measurement-request-count']} ")
-                return 1
+                        f" of {self._base_perf_config['measurement-request-count']} "
+                    )
+                return self.PA_FAIL
+
+        return self.PA_SUCCESS
+
+    def _execute_pa(self, env):
```
Comment (nv-braf) [inline model_analyzer/perf_analyzer/perf_analyzer.py:169]: Seems like this can be refactored into:

- Set cmd
- Set perf_analyzer_env
- actually execute (try: section)
- poll and check return code

File: model_analyzer/result/result_manager.py [hunk 4]
```diff
@@ -202,7 +202,7 @@ def create_tables(self,
                     for gpu_metric in gpu_metrics:
                         if gpu_metric.tag not in gpu_specific_metrics_from_measurements:
                             gpu_specific_metrics_from_measurements[
-                                gpu_metric.tag] = gpu_metric
+                                gpu_metric[0]] = gpu_metric
```
Comment (nv-braf) [inline model_analyzer/result/result_manager.py:205]: I don't understand why this change was needed (shouldn't you still need the .tag?)

File: model_analyzer/result/results.py [hunk 5]
```diff
@@ -40,88 +40,86 @@ def from_dict(cls, results_dict):
...
-                for key, measurement_dict in model_config_tuple_list[
+                for key, measurement_dict in run_config_tuple_list[
                         Results.MEASUREMENTS_INDEX].items():
                     run_config_measurement = RunConfigMeasurement.from_dict(
                         measurement_dict)
 
-                    results._add_run_config_measurement(model_name,
-                                                        model_config,
-                                                        model_config_name, key,
+                    results._add_run_config_measurement(models_name, run_config,
```
Comment (nv-braf) [inline model_analyzer/result/results.py:62]: Sort of like attorneys general. I guarantee I'll mistype this every time! 😄

File: model_analyzer/analyzer.py (hunk 1 above)
Comment (tgerdesnv) [reply model_analyzer/analyzer.py]: done

File: model_analyzer/analyzer.py (hunk 2 above)
Comment (tgerdesnv) [reply model_analyzer/analyzer.py]: done

File: model_analyzer/perf_analyzer/perf_analyzer.py (hunk 3 above)
Comment (tgerdesnv) [reply model_analyzer/perf_analyzer/perf_analyzer.py:170]: done

File: model_analyzer/result/result_manager.py (hunk 4 above)
Comment (tgerdesnv) [reply model_analyzer/result/result_manager.py]: good catch. I confirmed that [0] is incorrect. May have been a leftover from a bad merge between some of your changes?

File: model_analyzer/result/results.py [hunk 6]
```diff
@@ -170,6 +168,7 @@ def get_list_of_run_config_measurements(self):
 
     def get_model_measurements_dict(self, model_name):
         """
+        # FIXME is this description correct?
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/result/results.py]: fixed

File: model_analyzer/result/results.py [hunk 7]
```diff
@@ -190,77 +189,69 @@ def get_model_measurements_dict(self, model_name):
 
         return self._results[model_name]
 
-    def get_model_config_measurements_dict(self, model_name, model_config_name):
+    def get_model_variants_measurements_dict(self, models_name,
+                                             model_variants_name):
         """
-        Given a model name and model config name, return a dict where the
-        key is model configs and the values are list of all associated measurements
+        FIXME: is this description correct?
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/result/results.py]: fixed

File: model_analyzer/result/results.py [hunk 8]
```diff
@@ -190,77 +189,69 @@ def get_model_measurements_dict(self, model_name):
...
+        model_variant_name: str
 
         Returns
         -------
-        Tuple - (ModelConfig, list of RunConfigMeasurements)
+        Tuple - (RunConfig, list of RunConfigMeasurements)
             Tuple consisting of the model_config and a list of
             all associated measurement values
         """
+        # FIXME -- run config or model config here?
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/result/results.py]: fixed
